File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <iostream>
#include <vector>

using namespace std;

int DFS(const vector<int>& numbers, int target, int index, int current) {
    if (index == numbers.size()) {
        if (current == target) {
            return 1;
        } else {
            return 0;
        }
    }

    int plusResult = DFS(numbers, target, index + 1, current + numbers[index]);
    int minusResult = DFS(numbers, target, index + 1, current - numbers[index]);

    return plusResult + minusResult;
}

int solution(vector<int> numbers, int target) {
    return DFS(numbers, target, 0, 0);
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cs
using System;

public class Solution 
{
    int result = 0;
    int t = 0;
    public int solution(int[] numbers, int target) 
    {
        t = target;
        DFS(0, 0, numbers);
        return result;
    }
    
    void DFS(int idx, int sum, int[] numbers)
    {
        if (idx == numbers.Length)
        {
            if(sum == t)
            {
                result++;
            }
            return;
        }

        DFS(idx + 1, sum + numbers[idx], numbers); // 양수로 선택
        DFS(idx + 1, sum - numbers[idx], numbers); // 음수로 선택
    }
}